Namespace discovery proposal (#7534)
kiali/kiali: design/KEPS/namespace-discovery/proposal.md

# Motivation

> Currently, the Kiali Server and Operator have many different ways to configure which namespaces should be accessible and viewable by the Kiali UI. The configurations are complex and confusing to users. We want to make things more easily configurable and understandable while still maintaining the ability for Kiali to have either cluster-wide access or limited access to specific namespaces. We also want kiali to support soft multitenancy using the same mechanism as istio to make configuration more consistent with istio and easier for users.

Reviewer: We need to define what we mean by "soft" multitenancy somewhere in here. Maybe even define "multitenancy" to differentiate what we mean.

Oh, and missing capitalization causes my OCD to flare up :)
```suggestion
Currently, the Kiali Server and Operator have many different ways to configure which namespaces should be accessible and viewable by the Kiali UI. The configurations are complex and confusing to users. We want to make things more easily configurable and understandable while still maintaining the ability for Kiali to have either cluster-wide access or limited access to specific namespaces. We also want Kiali to support soft multitenancy using the same mechanism as Istio to make configuration more consistent with Istio and easier for users.
```

Author: "soft" in this case means kiali still has cluster wide access but limits what it shows / operates on to the namespaces selected by discovery selectors.

Cluster A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Cluster B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector cluster A:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Kiali discovery selector cluster B:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

The same rules for multiple revisions in a multi-cluster deployment would still apply.

Primary-remote deployments would behave the same as single cluster where the remote cluster is scoped to the primary controlplane's discovery selectors.

TODO: external controlplane

### Failure scenarios

Since the discovery selectors act as a way to separate multiple Kiali deployments from one another, when Kiali relies on autodiscovering the Istio discovery selectors and it cannot auto discover them, the Kiali server will fail to start and/or panic rather than potentially show out of scope data. In a multi-primary deployment, if one primary cannot be reached then that cluster will be considered inaccessible.

## Discovery Selector Kiali Config Option

> A new configuration option will be added that will mimic the Istio discovery selectors configuration option. Using the configuration option, users will be able to specify which namespaces Kiali should be scoped to. Having a config option within Kiali has the advantage of not relying on autodiscovery of the Istio controlplanes. In cases where Kiali temporarily cannot access the control plane configuration, Kiali does not need to fail since it will rely on its own discovery selectors to scope itself. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.

Reviewer: Question is default behavior - IF the user does not set the  Kiali DS, what is the default behavior? I assume it is, "Use the Istio DS." If Istio DS cannot be found, then do what? Abort with error, I assume.

If the user DOES specify Kiali DS, then I assume Istio DS is ignored entirely.

Reviewer 2: Do we need an option to use Istio DS *if possible*?

Reviewer: I don't think so, unless a nil Kiali DS can have more than one semantic. I assume you either set Kiali DS and ignore Istio DS, or you do NOT set Kiali DS and you use Istio DS.

When would you set Kiali DS AND STILL want to use Istio DS?
When would you NOT set Kiali DS AND NOT want to use Istio DS either?

Author: > Question is default behavior - IF the user does not set the Kiali DS, what is the default behavior? I assume it is, "Use the Istio DS." If Istio DS cannot be found, then do what? Abort with error, I assume.

Yes.

> If the user DOES specify Kiali DS, then I assume Istio DS is ignored entirely.

Yes.

> Do we need an option to use Istio DS if possible?

Something like "fallback to Kiali DS but try to use Istio DS first"? That would be useful but I also don't want to add more config options than we have to.

Reviewer: > Something like "fallback to Kiali DS but try to use Istio DS first"? 

Oh, that's an interesting one. Er... I agree I don't want more config than we have to. Is there a real use-case for this "try Istio DS first, even if Kiali DS is set"?

> Kiali would add a top level `discoverySelectors` config option where you can specify a `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to.

Reviewer: I don't think it should be top level - seems appropriate under "deployment" - where "accessible_namespaces" are today.

Author: I wasn't sure the best place for it. `Deployment` seems to be config options related to how Kiali is deployed. The other options in there are stuff like `InstanceName`, `Namespace`, etc. all things related to the actual Kiali deployment. Whereas this option doesn't have anything to do with how Kiali itself is deployed. I think `accessible_namespaces` made sense there because the operator would change how Kiali was deployed based on that field but that is what `cluster_wide_access` is now for. I'm happy to change this though.

Reviewer: Could go under "istio" since this is supposed to declaring the mesh namespaces, but I don't want to add one more thing there :)

`deployment` is still applicable in the sense that it will determine where the Roles will be deployed (just like accessible namespaces).

This just doesn't seem a large enough "thing" to warrant being its own top-level setting. And since it kinda replaces accessible_namespaces, seems like it will be less confusing to put it under the  same category.

I am not strongly of opinion either way though :)

Make sure its called "discovery_selectors" though (snake_case, not camelCase) to remain consistent with everything else. :)

# Motivation

Currently, the Kiali Server and Operator have many different ways to configure which namespaces should be accessible and viewable by the Kiali UI. The configurations are complex and confusing to users. We want to make things more easily configurable and understandable while still maintaining the ability for Kiali to have either cluster-wide access or limited access to specific namespaces. We also want Kiali to support soft multitenancy using the same mechanism as Istio to make configuration more consistent with Istio and easier for users.

## Goals

- Provide an easy way for users to configure Kiali UI to have access to specific namespaces.
- Provide a way for Kiali Server to have either cluster-wide access to all namespaces or limited access to a subset of namespaces.
- Remove the deprecated functionality that provided the old way of filtering and limiting namespace access.
- Enable soft multi-tenancy Kiali deployments where multiple Kialis with cluster wide access can be deployed to the same cluster and scoped to specific namespaces.

# Solution

> Provide a new list of "discovery selectors" in the same way that Istio itself has discovery selectors. Kiali will re-use the Istio Discovery Selectors it retreives from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors.

Reviewer: ````suggestion
Provide a new list of "discovery selectors" in the same way that Istio itself has discovery selectors. Kiali will re-use the Istio Discovery Selectors it retrieves from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors. It will support [Kubernetes set-based selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements).  For example:
```yaml
spec:
  deployment:
    discovery_selectors:
      match_labels:
        abc: def
        uvw: xyz
      match_expressions:
      - key: tier
        operator: In
        values: ["production"]
      - key: region
        operator: NotIn
        values: ["east"]
```
````

> Provide a new list of "discovery selectors" in the same way that Istio itself has discovery selectors. Kiali will re-use the Istio Discovery Selectors it retreive's from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors.

Reviewer: ```suggestion
Provide a new list of "discovery selectors" in the same way that Istio itself has discovery selectors. Kiali will re-use the Istio Discovery Selectors it retreives from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors.
```

## Istio Discovery Selectors

> Kiali already autodiscovers the istio control planes. Discovery selectors for the istio control plane can be found by looking at the control plane's `istio` configmap under `mesh.discoverySelectors`. Kiali will read this configmap and if discovery selectors have been set, Kiali will scope itself to the namespaces selected by the discovery selectors. In cases where there are multiple control planes, for example with revisioned upgrades, Kiali will read both discovery selectors and OR the two together scoping itself to the disjoint of the two sets. For example:

Reviewer: ```suggestion
Kiali already autodiscovers the Istio control planes. Discovery selectors for the Istio control plane can be found by looking at the control plane's `istio` configmap under `mesh.discoverySelectors`. Kiali will read this configmap and if discovery selectors have been set, Kiali will scope itself to the namespaces selected by the discovery selectors. In cases where there are multiple control planes, for example with revisioned upgrades, Kiali will read both discovery selectors and OR the two together scoping itself to the disjoint of the two sets. For example:
```

Control plane A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Control plane B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
  - matchLabels:
      team: api
```

> For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selector's defined on that cluster.

Reviewer: ```suggestion
For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selectors defined on that cluster.
```

Cluster A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Cluster B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector cluster A:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Kiali discovery selector cluster B:

```yaml

> discoverySelectors:
>   - matchLabels:
>       team: backend

Reviewer: Follow operator convention - snake_case:

```suggestion
discovery_selectors:
  - match_labels:
      team: backend
```

```

The same rules for multiple revisions in a multi-cluster deployment would still apply.

Primary-remote deployments would behave the same as single cluster where the remote cluster is scoped to the primary controlplane's discovery selectors.

> TODO: external controlplane

Reviewer 2: ```suggestion
TODO: external control plane
```

## Discovery Selector Kiali Config Option

A new configuration option will be added that will mimic the Istio `discoverySelectors` configuration option. Using the configuration option, users will be able to similarly scope Kiali to the defined set of namespaces. This configuration option would take precedence over the autodiscovered control plane discovery selectors. Having a config option within Kiali has the advantage of not relying on auto-discovery via the Istio control plane. Kiali does not need to fail if it cannot, even temporarily, access the control plane configuration, because it can rely on its own discovery selectors configuration. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.

Kiali would add a `deployment.discovery_selectors` config option where you can specify global discovery selectors or cluster specific overrides `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to. Cluster specific overrides will take precedence over the global option for that cluster.

```yaml

> discoverySelectors:
>   global:
>     - labelSelector:
>   cluster_name:
>     - labelSelector:

Reviewer: Follow operator convention - snake_case:
```suggestion
discovery_selectors:
  global:
    - label_selector:
  cluster_name:
    - label_selector:
```

```

For multi-primary with different discovery selectors:

```yaml

> discoverySelectors:
>   east:
>     - matchLabels:
>         team: backend
>   west:
>     - matchLabels:
>         team: api

Reviewer: Follow operator convention - snake_case:

```suggestion
discovery_selectors:
  east:
    - match_labels:
        team: backend
  west:
    - match_labels:
        team: api
```

```

## No cluster wide access

Kiali relies on `deployment.accessible_namespaces` when `deployment.cluster_wide_access = false` to know which namespaces the server should create a kube cache for. The operator also uses `deployment.accessible_namespaces` to create a `Role/RoleBinding` for the Kiali server's Service Account giving the Kiali server access to that namespace. In order for the operator to create RBAC rules for the Kiali server, the operator will need to reconcile the discovery selectors with namespaces that exist in the cluster. The operator can take what is set in `deployment.discovery_selectors` and set the Kiali server's `deployment.discovery_selectors` config field to be the result of this reconcilation. For example:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:

>       - matchLabels:

Reviewer: Follow operator convention - snake_case:

```suggestion
      - match_labels:
```

          team: backend
```

would be translated into the Kiali config as:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:

>       - matchLabels:
>           kubernetes.io/metadata.name: backend-app1
>       - matchLabels:

Reviewer: Follow operator convention - snake_case:

```suggestion
      - match_labels:
          kubernetes.io/metadata.name: backend-app1
      - match_labels:
```

>           kubernetes.io/metadata.name: backend-app2

Reviewer: I actually like this better than the "keep accessible_namespaces" idea - even though the latter would be easier to implement than the former :) We'd completely get rid of accessible_namespaces which avoids any confusion about whether the "old" accessible_namespaces still works or not and just keeps one way in the server to get the list of namespaces - through discovery selectors (and we already have code in the server that does that).

```

An alternative is to keep `deployment.accessible_namespaces` and make it an "internal" field that only the operator sets and not users.

```yaml
spec:
  deployment:
    accessible_namespaces:
      - backend-app1
      - backend-app2
    cluster_wide_access: false
    discovery_selectors:

>       - matchLabels:

Reviewer: Follow operator convention - snake_case:

```suggestion
      - match_labels:
```

    - match_labels:
        abc: def
        uvw: xyz
    - match_labels:
        org: accounting
      match_expressions:
      - key: tier
        operator: In
        values: ["production"]
      - key: region
        operator: NotIn
        values: ["east"]
    - match_expressions:
      - key: country
        operator: In
        values: ["us", "ca"]
```

## Istio Discovery Selectors

Kiali already autodiscovers the Istio control planes. Discovery selectors for the Istio control plane can be found by looking at the control plane's `istio` configmap under `mesh.discoverySelectors`. Kiali will read this configmap and if discovery selectors have been set, Kiali will scope itself to the namespaces selected by the discovery selectors. In cases where there are multiple control planes, for example with revisioned upgrades, Kiali will read both discovery selectors and OR the two together scoping itself to the disjoint of the two sets. For example:

Control plane A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Control plane B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector:

> ```yaml
> discovery_selectors:
>   - match_labels:
>       team: backend
>   - match_labels:
>       team: api

Reviewer: ```suggestion
```yaml
discovery_selectors:
  global:
  - match_labels:
      team: backend
  - match_labels:
      team: api
```

```

For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selectors defined on that cluster.

Cluster A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Cluster B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

> Kiali discovery selector cluster A:
> 
> ```yaml
> discovery_selectors:
>   - match_labels:
>       team: api
> ```
> 
> Kiali discovery selector cluster B:
> 
> ```yaml
> discovery_selectors:
>   - match_labels:
>       team: backend
> ```

Reviewer: ````suggestion
Kiali discovery selector cluster A:

```yaml
discovery_selectors:
  global:
  - match_labels:
      team: api
```

Kiali discovery selector cluster B:

```yaml
discovery_selectors:
  global:
  - match_labels:
      team: backend
```
````

```yaml
discovery_selectors:
  global:
  - <array of label selectors>
  cluster_name:
  - <array of label selectors>
```

For multi-primary with different discovery selectors:

```yaml
discovery_selectors:
  east:
    - match_labels:
        team: backend
  west:
    - match_labels:
        team: api
```

## No cluster wide access

Kiali relies on `deployment.accessible_namespaces` when `deployment.cluster_wide_access = false` to know which namespaces the server should create a kube cache for. The operator also uses `deployment.accessible_namespaces` to create a `Role/RoleBinding` for the Kiali server's Service Account giving the Kiali server access to that namespace. In order for the operator to create RBAC rules for the Kiali server, the operator will need to reconcile the discovery selectors with namespaces that exist in the cluster. The operator can take what is set in `deployment.discovery_selectors` and set the Kiali server's `deployment.discovery_selectors` config field to be the result of this reconcilation. For example:

```yaml
spec:
  deployment:
    cluster_wide_access: false

>     discovery_selectors:
>       - match_labels:
>           team: backend

Reviewer: ```suggestion
    discovery_selectors:
      global:
      - match_labels:
          team: backend
```

```

would be translated into the Kiali config as:

```yaml
spec:
  deployment:
    cluster_wide_access: false

>     discovery_selectors:
>       - match_labels:
>           kubernetes.io/metadata.name: backend-app1
>       - match_labels:
>           kubernetes.io/metadata.name: backend-app2

Reviewer: ```suggestion
    discovery_selectors:
      global:
      - match_labels:
          kubernetes.io/metadata.name: backend-app1
      - match_labels:
          kubernetes.io/metadata.name: backend-app2
```

```

An alternative is to keep `deployment.accessible_namespaces` and make it an "internal" field that only the operator sets and not users.

```yaml
spec:
  deployment:
    accessible_namespaces:
      - backend-app1
      - backend-app2
    cluster_wide_access: false

>     discovery_selectors:
>       - match_labels:
>           team: backend

Reviewer: ```suggestion
    discovery_selectors:
      global:
      - match_labels:
          team: backend
```

- Provide a way for Kiali Server to have either cluster-wide access to all namespaces or limited access to a subset of namespaces.
- Remove the deprecated functionality that provided the old way of filtering and limiting namespace access.
- Enable soft multi-tenancy Kiali deployments where multiple Kialis with cluster wide access can be deployed to the same cluster and scoped to specific namespaces.

# Solution

Provide a new list of "discovery selectors" in the same way that [Istio itself has discovery selectors](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig). Kiali will re-use the Istio Discovery Selectors it retrieves from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors. It will support an array of [Kubernetes set-based selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) just like Istio Discovery Selectors.  For example:
```yaml
spec:
  deployment:

>     discovery_selectors:
>     - match_labels:
>         abc: def
>         uvw: xyz
>     - match_labels:
>         org: accounting
>       match_expressions:
>       - key: tier
>         operator: In
>         values: ["production"]
>       - key: region
>         operator: NotIn
>         values: ["east"]
>     - match_expressions:
>       - key: country
>         operator: In
>         values: ["us", "ca"]

Reviewer: ```suggestion
    discovery_selectors:
      global:
      - match_labels:
          abc: def
          uvw: xyz
      - match_labels:
          org: accounting
        match_expressions:
        - key: tier
          operator: In
          values: ["production"]
        - key: region
          operator: NotIn
          values: ["east"]
      - match_expressions:
        - key: country
          operator: In
          values: ["us", "ca"]
```

- Provide a way for Kiali Server to have either cluster-wide access to all namespaces or limited access to a subset of namespaces.
- Remove the deprecated functionality that provided the old way of filtering and limiting namespace access.
- Enable soft multi-tenancy Kiali deployments where multiple Kialis with cluster wide access can be deployed to the same cluster and scoped to specific namespaces.

# Solution

Provide a new list of "discovery selectors" in the same way that [Istio itself has discovery selectors](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig). Kiali will re-use the Istio Discovery Selectors it retrieves from the control plane's `istio` configmap. Kiali will also have a new config option `deployment.discovery_selectors` that, when specified, will be used instead of the control plane's Istio Discovery Selectors. It will support an array of [Kubernetes set-based selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) just like Istio Discovery Selectors.  For example:
```yaml
spec:
  deployment:
    discovery_selectors:

>       global:

Reviewer: ```suggestion
      default:
```

      - match_labels:
          org: accounting
        match_expressions:
        - key: tier
          operator: In
          values: ["production"]
        - key: region
          operator: NotIn
          values: ["east"]
      - match_expressions:
        - key: country
          operator: In
          values: ["us", "ca"]
```

## Istio Discovery Selectors

Kiali already autodiscovers the Istio control planes. Discovery selectors for the Istio control plane can be found by looking at the control plane's `istio` configmap under `mesh.discoverySelectors`. Kiali will read this configmap and if discovery selectors have been set, Kiali will scope itself to the namespaces selected by the discovery selectors. In cases where there are multiple control planes, for example with revisioned upgrades, Kiali will read both discovery selectors and OR the two together scoping itself to the disjoint of the two sets. For example:

Control plane A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Control plane B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector:

```yaml
discovery_selectors:

>   global:

Reviewer: ```suggestion
  default:
```

  - match_labels:
      team: backend
  - match_labels:
      team: api
```

For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selectors defined on that cluster.

Cluster A discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: api
```

Cluster B discovery selector:

```yaml
discoverySelectors:
  - matchLabels:
      team: backend
```

Kiali discovery selector cluster A:

```yaml
discovery_selectors:

>   global:

Reviewer: ```suggestion
  default:
```

  - match_labels:
      team: api
```

Kiali discovery selector cluster B:

```yaml
discovery_selectors:

>   global:

Reviewer: ```suggestion
  default:
```

## Discovery Selector Kiali Config Option

A new configuration option will be added that will mimic the Istio `discoverySelectors` configuration option. Using the configuration option, users will be able to similarly scope Kiali to the defined set of namespaces. This configuration option would take precedence over the autodiscovered control plane discovery selectors. Having a config option within Kiali has the advantage of not relying on auto-discovery via the Istio control plane. Kiali does not need to fail if it cannot, even temporarily, access the control plane configuration, because it can rely on its own discovery selectors configuration. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.

> Kiali would add a `deployment.discovery_selectors` config option where you can specify global discovery selectors or cluster specific overrides `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to. Cluster specific overrides will take precedence over the global option for that cluster.

Reviewer: ```suggestion
Kiali would add a `deployment.discovery_selectors` config option where you can specify a default set of discovery selectors and additional cluster-specific overrides `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to. Cluster specific overrides will take precedence over the defaults.
```

```yaml
discovery_selectors:

>   global:

Reviewer: ```suggestion
  default:
```

  - <array of label selectors>
  cluster_name:
  - <array of label selectors>
```

For multi-primary with different discovery selectors:

```yaml
discovery_selectors:
  east:
    - match_labels:
        team: backend
  west:
    - match_labels:
        team: api
```

## No cluster wide access

Kiali relies on `deployment.accessible_namespaces` when `deployment.cluster_wide_access = false` to know which namespaces the server should create a kube cache for. The operator also uses `deployment.accessible_namespaces` to create a `Role/RoleBinding` for the Kiali server's Service Account giving the Kiali server access to that namespace. In order for the operator to create RBAC rules for the Kiali server, the operator will need to reconcile the discovery selectors with namespaces that exist in the cluster. The operator can take what is set in `deployment.discovery_selectors` and set the Kiali server's `deployment.discovery_selectors` config field to be the result of this reconcilation. For example:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:

>       global:

Reviewer: ```suggestion
      default:
```

      - match_labels:
          team: backend
```

would be translated into the Kiali config as:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:

>       global:

Reviewer: ```suggestion
      default:
```

      - match_labels:
          kubernetes.io/metadata.name: backend-app1
      - match_labels:
          kubernetes.io/metadata.name: backend-app2
```

An alternative is to keep `deployment.accessible_namespaces` and make it an "internal" field that only the operator sets and not users.

```yaml
spec:
  deployment:
    accessible_namespaces:
      - backend-app1
      - backend-app2
    cluster_wide_access: false
    discovery_selectors:

>       global:

Reviewer: ```suggestion
      default:
```

Since the discovery selectors act as a way to separate multiple Kiali deployments from one another, when Kiali relies on autodiscovering the Istio discovery selectors and it cannot auto discover them, the Kiali server will fail to start and/or panic rather than potentially show out of scope data. In a multi-primary deployment, if one primary cannot be reached then that cluster will be considered inaccessible.

## Discovery Selector Kiali Config Option

> A new configuration option will be added that will mimic the Istio discovery selectors configuration option. Using the configuration option, users will be able to specify which namespaces Kiali should be scoped to. Having a config option within Kiali has the advantage of not relying on autodiscovery of the Istio controlplanes. In cases where Kiali temporarily cannot access the control plane configuration, Kiali does not need to fail since it will rely on its own discovery selectors to scope itself. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.

Reviewer 2: ```suggestion
A new configuration option will be added that will mimic the Istio `discoverySelectors` configuration option. Using the configuration option, users will be able to similarly scope Kiali to the defined set of namespaces. Having a config option within Kiali has the advantage of not relying on auto-discovery via the Istio control plane. Kiali does not need to fail if it cannot, even temporarily, access the control plane configuration, because it can rely on its own discovery selectors configuration. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.
```

> Kiali would add a top level `discoverySelectors` config option where you can specify a `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to.

Reviewer 2: I think perhaps we should have a global discoverySelector and only cluster overrides.  This would allow a simple definition that can apply to all clusters, which makes things easy for single-cluster and scenarios where all clusters use the same thing.

Author: I like that option.

Reviewer: Is it a valid edge case that clusters can have different DS? I'm sure Istio allows it (because, one control plane probably doesn't care about another control plane's DS) but is it a valid thing to do (do people do that?). 

And I thought Kiali assumes "namespace sameness" in multi-cluster?

Author: > Is it a valid edge case that clusters can have different DS? I'm sure Istio allows it (because, one control plane probably doesn't care about another control plane's DS) but is it a valid thing to do (do people do that?).

I'd say yes. It's perfectly valid to have namespaces that exist in one cluster but don't exist on another.

> And I thought Kiali assumes "namespace sameness" in multi-cluster?

You can't assume that the same namespace will exist on both clusters and that the two clusters will be exactly the same. The actual namespace object only exists inside a single cluster even if istio logically treats an app deployed in the same namespace across both clusters as the same app.

Author: > We could probably hold off on the cluster-specific impl until it's proven we need it.

I'd rather give users the ability to specify this now or at the very least keep the structure the same where we have a `global:` setting so that we can easily add cluster specific overrides later.

Reviewer: I'll add "global:" to all the examples where there is none here in this document. The one issue I'm wrestling with is how will the operator know the home cluster name so it can do its own override when processing the selectors (because it needs to do this in order to create the Roles in the proper places).

For example, suppose we have this, for a hard multi-tenancy setup:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:
      global:
      - match_labels:
          foo: howdy
      cluster-a:
      - match_labels:
          foo: hello
      cluster-b:
      - match_labels:
          foo: wassup
```

Suppose Kiali's home cluster is `cluster-b` -- how does the operator know that? Because the operator will need to know that, since it will need to know that `cluster-b` is overriding `global`.

(BTW: I'm still not clear on what it means for one cluster's selectors to override the global selectors - I assume it means the cluster-specific one just completely overrides the global, where I could have a dozen more clusters, but because they aren't specified in `discovery_selectors`, then just the `global` ones are used, but for cluster-a and cluster-b, the entire array of selectors under global are ignored.)

Reviewer: Hmm.. I think I might have an answer to my question above about "how does the operator know the home cluster?". The operator can look at [spec.clustering.clusters[*].name](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.clustering.clusters[*].name) - those are all the remote cluster names. If there is an override cluster section under "discovery_selectors" that does NOT appear in that list of remote cluster names, it must be the home cluster. This assumes every remote cluster is specified in that remote cluster section (even those not accessible).

Reviewer 2: Instead of `global` maybe it should be `default`.  I would expect if there is a cluster-override it would not be additive, it would be a replacement.

Reviewer: We could also say the "global" (or "default") simply IS the home cluster. So maybe instead of global or default we call it "home" or "kiali-home" or "local" (we want it unique enough for it not to conflict with any cluster name a user might want to use - so "kiali-home" seems like something people won't use, but "local" and "home" seem more useful cluster names for users to want to use).

So you would define the kiali home cluster and it is used as the default (which, when you think about it, is likely going to be the common global selectors anyway - at least for the majority of use cases. So that seems like a viable "default" for "global" selectors).

So instead of:

```yaml
deployment_selectors:
  global:
    foo: hello
  cluster-a:
    foo: howdy
```

it would be:
```yaml
deployment_selectors:
  kiali-home:
    foo: hello
  cluster-a:
    foo: howdy
```

That seems to make sense to me

Reviewer: I also think that we should change "global" for no other reason than that seems like it could potentially be a cluster name that a user might want ("global" as in "worldwide", then you have different regions like "emea", "na", etc).

I think it should be "kiali-home" because (a) it is more descriptive and (b) likely won't clash with any name a user might want to use.

Reviewer: we ended up just going with "default" for now.

Author: > The one issue I'm wrestling with is how will the operator know the home cluster name so it can do its own override when processing the selectors (because it needs to do this in order to create the Roles in the proper places)

The operator should not be cluster aware. Today iirc the "hard multi-tenancy mode" of `accessible_namespaces` does not work for multi-cluster. Unless that's changing the operator would just configure the roles for the cluster that the operator is deployed on and not try to do anything cluster related.

Author: > I also think that we should change "global" for no other reason than that seems like it could potentially be a cluster name that a user might want ("global" as in "worldwide", then you have different regions like "emea", "na", etc).

Hmmm that's a good point and any name will eventually conflict so imo we should just change the structure to something like:
```
deployment_selectors:
  global (or default):
    foo: hello
  overrides:
    <cluster-name>:
      foo: howdy
```
Merging global selectors with override selectors would be more complicated so I think the simplest heuristic would be to only use what is specified in `overrides` if you have specified anything.

Reviewer 2: I have no problem with adding an `overrides`.  We already preferred `default`, and `overrides` works well with `default`.   I'm -1, though on ignoring `default` if

Author: > I think you mean that if there is a cluster-specific override that you would ignore default for that cluster, and any cluster not included in overrides would get the defaults.

Yes exactly

Reviewer: OK. let's change this KEP to refer to this extra level "overrides".
